test/e2e/route_metrics_test.go

```diff
@@ -0,0 +1,393 @@
+		if err := kclient.Get(context.TODO(), routeFooLabelName, routeFooLabel); err != nil {
+			t.Fatalf("failed to get route resource: %v", err)
+		}
+		routeFooLabel.Labels = map[string]string{
+			"type": incorrectLabel,
+		}
+		if err := kclient.Update(context.TODO(), routeFooLabel); err != nil {
+			t.Fatalf("failed to update route: %v", err)
+		}
+
+		// Wait for metrics to be updated to 0 as the Route will get un-admitted by the IC.
+		if err := waitForRouteMetricsAddorUpdate(t, prometheusClient, ic.Name, 0); err != nil {
+			t.Fatalf("failed to fetch expected metrics: %v", err)
+		}
+
+		// Update the label of the route so that it gets admitted to the Ingress Controller again.
+		if err := kclient.Get(context.TODO(), routeFooLabelName, routeFooLabel); err != nil {
+			t.Fatalf("failed to get route resource: %v", err)
+		}
+		routeFooLabel.Labels = map[string]string{
+			"type": correctLabel,
+		}
+		if err := kclient.Update(context.TODO(), routeFooLabel); err != nil {
+			t.Fatalf("failed to update route: %v", err)
+		}
+
+		// Wait for metrics to be updated to 1 as the Route will get admitted by the IC again.
+		if err := waitForRouteMetricsAddorUpdate(t, prometheusClient, ic.Name, 1); err != nil {
+			t.Fatalf("failed to fetch expected metrics: %v", err)
+		}
+	}
+
+	// Delete the Route routeFooLabel.
+	func() {
+		if err := kclient.Delete(context.TODO(), routeFooLabel); err != nil {
+			if apierrors.IsNotFound(err) {
+				return
+			} else {
+				t.Fatalf("failed to delete route %s: %v", routeFooLabelName, err)
+			}
+		}
+	}()
+
+	// Wait for metrics to be updated to zero as the admitted Route is deleted.
+	if err := waitForRouteMetricsAddorUpdate(t, prometheusClient, ic.Name, 0); err != nil {
+		t.Fatalf("failed to fetch expected metrics: %v", err)
+	}
+
+	// Delete the IC.
+	assertIngressControllerDeleted(t, kclient, ic)
+
+	// Wait for metrics corresponding to the IC to be deleted.
+	if err := waitForRouteMetricsDelete(t, prometheusClient, ic.Name); err != nil {
+		t.Fatalf("failed to fetch expected metrics: %v", err)
+	}
+}
+
+// waitForRouteMetricsAddorUpdate waits for the metrics for the corresponding shard to be added or updated to the expected value.
+func waitForRouteMetricsAddorUpdate(t *testing.T, prometheusClient prometheusv1.API, shardName string, value int) error {
+	if err := wait.PollImmediate(1*time.Second, 2*time.Minute, func() (bool, error) {
+		result, _, err := prometheusClient.Query(context.TODO(), fmt.Sprintf(`route_metrics_controller_routes_per_shard{name="%s"}`, shardName), time.Now())
+		if err != nil {
+			t.Logf("failed to fetch metrics: %v", err)
+			return false, nil
+		}
+
+		// Check if fetched metrics is of Vector type.
+		vec, ok := result.(model.Vector)
+		if !ok {
+			return ok, nil
+		}
+
+		// Check if length of returned metric Vector is zero.
+		if len(vec) == 0 {
+			return false, nil
+		}
+
+		// Check if metrics is updated.
+		if !vec[0].Value.Equal(model.SampleValue(float64(value))) {
```

Comment:
I wonder if this introduces a race condition where the longer it takes to get to this point, the more likely the positional index in the vector might not be the 0th.  Can you change the query so you only get one value back, and then you don't have to bother with the vector?  Maybe add a time range to the query that only gets the very last sample?  Something like`route_metrics_controller_routes_per_shard{name="%s"} [30s]` ?

Reply:
The query, `route_metrics_controller_routes_per_shard{name="%s"}`, is an `instant vector selector` type (https://prometheus.io/docs/prometheus/latest/querying/basics/#instant-vector-selectors). As we are filtering by the label `name` for the particular shard it will only return the time series of that particular `shard` for the `current timestamp`. As the shard names are unique, the query will return a vector of size 1 if the metric for the `shard` exists and it will always be available at the 0th index.

The query, `route_metrics_controller_routes_per_shard{name="%s"}[30s]`, is a `range vector selector` type (https://prometheus.io/docs/prometheus/latest/querying/basics/#range-vector-selectors). This will return a range of values for each time series returned by the query for the specified duration. That means, the query will return all the values that were assigned to the metric for the particular `shard` in the last 30s. This query will lead to issues of which position has the latest value and might lead to incorrect e2e test results. I ran the query for the `default` shard and got the following result:

![Screenshot from 2022-10-10 16-14-16](https://user-images.githubusercontent.com/5279033/194852809-85750352-8272-47c4-9d1d-9589e74e2886.png)

It can be seen that 2 values are being returned as the value was changed from `10` to `9` in the last 30s for the `default` shard.

Reply:
Well said.  Thanks for the explanation.